File: libs/monobitmap-manager/src/main/kotlin/mono/bitmap/manager/MonoBitmapManager.kt
package mono.bitmap.manager

import mono.bitmap.manager.factory.LineBitmapFactory
import mono.bitmap.manager.factory.RectangleBitmapFactory
import mono.bitmap.manager.factory.TextBitmapFactory
import mono.graphics.bitmap.MonoBitmap
import mono.shape.shape.AbstractShape
import mono.shape.shape.Group
import mono.shape.shape.Line
import mono.shape.shape.MockShape
import mono.shape.shape.Rectangle
import mono.shape.shape.Text

/**
 * A model class which manages and caches bitmap of shapes.
 * Cache-hit when both id and version of the shape valid in the cache, otherwise, cache-miss.
 */
class MonoBitmapManager {
    private val idToBitmapMap: MutableMap<String, VersionizedBitmap> = mutableMapOf()

    fun getBitmap(shape: AbstractShape): MonoBitmap? {
        val cachedBitmap = getCacheBitmap(shape)
        if (cachedBitmap != null) {
            return cachedBitmap
        }

        val bitmap = when (shape) {
            is Rectangle -> RectangleBitmapFactory.toBitmap(
                shape.bound.size,
                shape.extra
            )
            is Text -> TextBitmapFactory.toBitmap(
                shape.bound.size,
                shape.renderableText.getRenderableText(),
                shape.extra,
                shape.isTextEditing
            )
            is Line -> LineBitmapFactory.toBitmap(
                shape.reducedJoinPoints,
                shape.extra,
            )

            is Group -> null // No draw group since it change very frequently.
            is MockShape -> null // Only for testing.
        } ?: return null
        idToBitmapMap[shape.id] = VersionizedBitmap(shape.versionCode, bitmap)
        return bitmap
    }

    private fun getCacheBitmap(shape: AbstractShape): MonoBitmap? =
        idToBitmapMap[shape.id]?.takeIf { it.versionCode == shape.versionCode }?.bitmap

    private class VersionizedBitmap(val versionCode: Int, val bitmap: MonoBitmap)
}

File: monosketch-svelte/src/lib/mono/monobitmap/manager/mono-bitmap-manager.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import { Group } from "$mono/shape/shape/group";
import { Line } from "$mono/shape/shape/line";
import { MockShape } from "$mono/shape/shape/mock-shape";
import { Rectangle } from "$mono/shape/shape/rectangle";
import { Text } from "$mono/shape/shape/text";
import { LineBitmapFactory } from './line-bitmap-factory';
import { RectangleBitmapFactory } from './rectangle-bitmap-factory';
import { TextBitmapFactory } from './text-bitmap-factory';
import { MonoBitmap } from '$mono/monobitmap/bitmap/monobitmap';
import Bitmap = MonoBitmap.Bitmap;

/**
 * A model class which manages and caches bitmap of shapes.
 * Cache-hit when both id and version of the shape valid in the cache, otherwise, cache-miss.
 */
export class MonoBitmapManager {
    private idToBitmapMap: Map<string, VersionizedBitmap> = new Map();

    getBitmap(shape: AbstractShape): MonoBitmap.Bitmap | null {
        const cachedBitmap = this.getCacheBitmap(shape);
        if (cachedBitmap) {
            return cachedBitmap;
        }

        let bitmap = this.createBitmap(shape);

        if (bitmap) {
            this.idToBitmapMap.set(shape.id, { versionCode: shape.versionCode, bitmap });
        }
        return bitmap;
    }

    private createBitmap(shape: AbstractShape): Bitmap | null {
        if (shape instanceof Rectangle) {
            return RectangleBitmapFactory.toBitmap(shape.bound.size, shape.extra);
        } else if (shape instanceof Text) {
            return TextBitmapFactory.toBitmap(
                shape.bound.size,
                shape.renderableText.getRenderableText(),
                shape.extra,
                shape.isTextEditing,
            );
        } else if (shape instanceof Line) {
            return LineBitmapFactory.toBitmap(shape.reducedJoinPoints, shape.extra);
        } else if (shape instanceof Group || shape instanceof MockShape) {
            // No draw group since it changes very frequently.
            // No draw mock shape since it is only for testing.
            return null;
        } else {
            throw new Error(`Unsupported shape type: ${shape.constructor.name}`);
        }
    }

    private getCacheBitmap(shape: AbstractShape): MonoBitmap.Bitmap | null {
        const versionizedBitmap = this.idToBitmapMap.get(shape.id);
        return versionizedBitmap && versionizedBitmap.versionCode === shape.versionCode
            ? versionizedBitmap.bitmap
            : null;
    }
}

interface VersionizedBitmap {
    versionCode: number;
    bitmap: MonoBitmap.Bitmap;
}